### PR TITLE
Add read and write timeouts

### PR DESCRIPTION
**Pull request by Choraden**
This patch makes the proxy more configurable when it comes to waiting for read/write.
Global timeout is now exchanged for
 - ReadTimeout - maximum duration for reading the entire request, including the body
 - ReadHeaderTimeout - maximum duration for reading request headers
 - WriteTimeout - maximum duration for writing the response

### CODE REVIEW

File: proxy.go [hunk 1]
```diff
@@ -63,9 +63,33 @@ type Proxy struct {
 	// ErrorResponse specifies a custom error HTTP response to send when a proxying error occurs.
 	ErrorResponse func(req *http.Request, err error) *http.Response
 
+	// ReadTimeout is the maximum duration for reading the entire
```
Comment (mmatczuk) [inline proxy.go:68]: For ascetics could we add whitespace for the parameters above to?

File: proxy.go [hunk 2]
```diff
@@ -126,11 +149,6 @@ func (p *Proxy) SetDownstreamProxyFunc(f func(*http.Request) (*url.URL, error))
 	}
 }
 
-// SetTimeout sets the request timeout of the proxy.
-func (p *Proxy) SetTimeout(timeout time.Duration) {
```
Comment (mmatczuk) [inline proxy.go:130]: Let's leave that for the compatibility it would set both read and write t/o.

File: proxy_test.go [hunk 3]
```diff
@@ -1462,3 +1463,57 @@ func TestRacyClose(t *testing.T) {
...
+
+func TestReadHeaderTimeout(t *testing.T) {
+	t.Parallel()
+
+	expectedError := func(err error) bool {
+		if neterr, ok := err.(net.Error); ok && neterr.Timeout() {
+			return true
+		}
+
+		switch err {
```
Comment (mmatczuk) [inline proxy_test.go]: So what do we expect exactly?

Comment (Choraden) [reply proxy_test.go]: I find it confusing that some times I would get io.UnexpectedEOF the other time "connection reset by peer", so I decided to check for all these errors

Comment (Choraden) [reply proxy_test.go]: n/m, I found a workaround

File: proxy.go [hunk 4]
```diff
@@ -290,7 +313,30 @@ var (
...
 func (p *Proxy) readRequest(ctx *Context, conn net.Conn, brw *bufio.ReadWriter) (req *http.Request, err error) {
+	var (
+		wholeReqDeadline time.Time // or zero if none
+		hdrDeadline      time.Time // or zero if none
+	)
+	t0 := time.Now()
+	if d := p.readHeaderTimeout(); d > 0 {
+		hdrDeadline = t0.Add(d)
+	}
+	if p.ReadTimeout > 0 {
```
Comment (mmatczuk) [inline proxy.go]: It looks nicer in the original version to me, why change it.

```
	if d := c.server.readHeaderTimeout(); d > 0 {
		hdrDeadline = t0.Add(d)
	}
	if d := c.server.ReadTimeout; d > 0 {
		wholeReqDeadline = t0.Add(d)
	}
```

File: proxy.go [hunk 5]
```diff
@@ -309,6 +355,13 @@ func (p *Proxy) readRequest(ctx *Context, conn net.Conn, brw *bufio.ReadWriter)
 		}
 	}
 
+	// Adjust the read deadline if necessary.
```
Comment (mmatczuk) [inline proxy.go:358]: Nice trick.

File: proxy.go [hunk 6]
```diff
@@ -585,6 +638,12 @@ func (p *Proxy) handle(ctx *Context, conn net.Conn, brw *bufio.ReadWriter) error
 		}
 	}
 
+	if p.WriteTimeout != 0 {
```
Comment (mmatczuk) [inline proxy.go]: Maybe `> 0` ?

**Review by mmatczuk — APPROVED** [review]

**Pull request merged**